kitsune/middleware.py

```diff
@@ -0,0 +1,199 @@
+
+
+class LithiumRedirectionMiddleware():
+    """
+    Redirect requests for user profiles from /<username> to
+    /u/<username> to avoid breaking profile urls with the new url
+    schema.
+
+    """
+
+    meta_regex = re.compile('<meta content="(.*?)" property="(.*?)"/>')
+    product_pages = {
+        u'Mozilla-Hilfe-Deutsch': '/de/?cache=no',
+        u'Soporte-de-Mozilla-Español': '/es/?cache=no',
+        u'Magyar': '/hu/?cache=no',
+        u'Pomoc-Mozilli-polski': '/pl/?cache=no',
+        u'Podpora-Mozilly-slovenčina': '/sk/?cache=no',
+        u'Поддержка-Mozilla-Русский': '/ru/?cache=no',
+        u'Podpora-Mozilly-Čeština': '/cs/?cache=no',
+        u'Mozilla-Support-Ελληνικά': '/el/?cache=no',
+        u'Mozilla-tuki-suomi': '/fi/?cache=no',
+        u'Dukungan-Mozilla-Bahasa': '/hi-IN/?cache=no',
+        u'Suporte-Mozilla-Português-do': '/pt-br/?cache=no',
+        u'Mozilla-Destek-Türkçe': '/tr/?cache=no',
+        u'Підтримка-Mozilla-Українська': '/uk/?cache=no',
+        u'Mozilla-Support-Dansk': '/da/?cache=no',
+        u'Mozilla-Support-English': '/en/?cache=no',
+        u'Assistance-de-Mozilla-Français': '/fr/?cache=no',
+        u'Supporto-Mozilla-Italiano': '/it/?cache=no',
+        u'Mozilla-Support-Nederlands': '/nl/?cache=no',
+        u'Apoio-da-Mozilla-Português': '/pt/?cache=no',
+        u'Mozilla-Support-Svenska': '/sv/?cache=no',
+        'Mozilla-\xe3\x82\xb5\xe3\x83\x9d\xe3\x83\xbc\xe3\x83\x88-\xe6\x97\xa5\xe6\x9c\xac\xe8\xaa\x9e': '/ja/?cache=no',
+        'Mozilla-\xe6\x8a\x80\xe8\xa1\x93\xe6\x94\xaf\xe6\x8f\xb4-\xe6\xad\xa3\xe9\xab\x94\xe4\xb8\xad\xe6\x96\x87-\xe7\xb9\x81\xe9\xab\x94': '/zh-TW/?cache=no',
+        'Mozilla-\xeb\x8f\x84\xec\x9b\x80\xeb\xa7\x90-\xed\x95\x9c\xea\xb5\xad\xec\x96\xb4': '/ko/?cache=no',
+        'Mozilla-\xe6\x8a\x80\xe6\x9c\xaf\xe6\x94\xaf\xe6\x8c\x81-\xe4\xb8\xad\xe6\x96\x87-\xe7\xae\x80\xe4\xbd\x93': '/zh-CN/?cache=no',
+        '\xd8\xaf\xd8\xb9\xd9\x85-\xd9\x81\xd9\x8e\xd9\x8a\xd9\x8e\xd8\xb1\xd9\x81\xd9\x8f\xd9\x83\xd8\xb3-\xd8\xb9\xd8\xb1\xd8\xa8\xd9\x8a': '/ar/?cache=no',
+    }
+
+    extra_pages = {
+        '/t5/How-To/Cookies/ta-p/16348': '/kb/cookies-information-websites-store-on-your-computer',
+        '/t5/Fix-slowness-crashing-error/What-does-quot-Your-connection-is-not-secure-quot-mean/ta-p/30354': '/kb/what-does-your-connection-is-not-secure-mean',
+        '/t5/Firefox-for-Android/ct-p/Firefox-Android': '/products/mobile',
+    }
+
+    def process_response(self, request, response):
+        if (response.status_code != 404):
+            return response
+
+        urlparts = request.path_info.split('/')
+        try:
+            if not urlparts[1].startswith('t5'):
+                return response
+        except KeyError:
+            return response
+
+
+        # Oh boy, Lithium URL. Let's see if we can turn things around.
+
+        # Cache first
+        cache_key = hashlib.sha1(request.path_info.encode('utf-8')).hexdigest()
```

Comment:
It's not a big deal, but i would probably be better not to use sha1 anymore, as a general rule.